reactive/docker.py

```diff
@@ -257,55 +254,93 @@ def install_from_upstream_apt():
 def install_from_nvidia_apt():
     ''' Install cuda docker from the nvidia apt repository. '''
     status_set('maintenance', 'Installing docker-engine from Nvidia PPA.')
-
     # Get the server and key in the apt-key management tool.
-    ksvr = 'hkp://p80.pool.sks-keyservers.net:80'
     for key in ["C95B321B61E88C1809C4F759DDCAE044F796ECB0",
                 "9DC858229FC7DD38854AE2D88D81803C0EBFCD88"]:
-        cmd = 'apt-key adv --keyserver {0} --recv-keys {1}'.format(ksvr, key)
-        check_call(split(cmd))
+        add_apt_key(key)
 
     # Get the package architecture (amd64), not the machine hardware (x86_64)
-    arch = check_output(split('dpkg --print-architecture'))
-    arch = arch.decode('utf-8').rstrip()
+    architecture = arch()
     # Get the lsb information as a dictionary.
     lsb = host.lsb_release()
     code = lsb['DISTRIB_CODENAME']
     rel = lsb['DISTRIB_RELEASE']
+    ubuntu = str(lsb['DISTRIB_ID']).lower()
     dockurl = "https://download.docker.com/linux/ubuntu"
     nvidurl = 'https://nvidia.github.io'
     repo = 'stable'
 
     deb = list()
-    deb.append('deb [arch={0}] {1} {2} {3}'.format(arch, dockurl, code, repo))
+    deb.append('deb [arch={0}] {1} {2} {3}'.format(architecture,
+                                                   dockurl, code, repo))
     for i in ['libnvidia-container',
               'nvidia-container-runtime',
               'nvidia-docker']:
-        deb.append('deb {0}/{1}/ubuntu{2}/{3} /'.format(nvidurl, i, rel, arch))
+        deb.append('deb {0}/{1}/ubuntu{2}/{3} /'.format(nvidurl,
+                                                        i, rel, architecture))
+
+    write_docker_sources(deb)
+
+    install_cuda_drivers_repo(architecture, rel, ubuntu)
+
+    apt_update(fatal=True)
+    # actually install the required packages docker-ce nvidia-docker2
+
+    docker_ce = hookenv.config('docker-ce-package')
+    nvidia_docker2 = hookenv.config('nvidia-docker-package')
+    nv_container_runtime = hookenv.config('nvidia-container-runtime-package')
+    apt_install(['cuda-drivers', docker_ce, nvidia_docker2,
+                 nv_container_runtime], fatal=True)
+
 
+def install_cuda_drivers_repo(architecture, rel, ubuntu):
+    ''' Install cuda drivers this is xenial only.
+     We want to install cuda-drivers only this means that the
+     cuda version plays no role. Any repo will do. '''
+    architecture = arch()
+    # Get the lsb information as a dictionary.
+    lsb = host.lsb_release()
+    rel = lsb['DISTRIB_RELEASE']
+    ubuntu = str(lsb['DISTRIB_ID']).lower()
```

Comment:
We set this variable, pass it in and then smash it?

Reply:
You are right, failed method extraction.